src/openlcb/EventIdentifyGlobal.hxx

```diff
@@ -0,0 +1,153 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @file EventIdentifyGlobal.hxx
+ *
+ * Implements the necessary logic to produce EventIdentifyGlobal messages.
+ *
+ * @author Stuart Baker
+ * @date 4 April 2021
+ */
+
+#ifndef _OPENLCB_EVENTIDENTIFYGLOBAL_HXX_
+#define _OPENLCB_EVENTIDENTIFYGLOBAL_HXX_
+
+#include "openlcb/If.hxx"
+#include "utils/Atomic.hxx"
+
+namespace openlcb
+{
+
+/// Helper object for producing an Event Identify Global message. Once armed,
+/// will produce an Event Identify Global message on the node's interface
+/// at a pseudo random time between 1.500 and 2.011 seconds in the future. The
+/// randomness comes from the 9 least significant bits of the Node ID and aids
+/// in reducing the likeliness of multiple nodes of the same design producing
+/// the same Event Identify Global message unnecessarily.
+class EventIdentifyGlobal : public StateFlowBase, private Atomic
+{
+public:
+    /// Constructor.
+    /// @param node Node ID associated with this object
+    EventIdentifyGlobal(Node *node)
+        : StateFlowBase(node->iface())
+        , eventIdentifyGlobalHandler_(
+              this, &EventIdentifyGlobal::event_identify_global)
+        , timer_(this)
+        , node_(node)
+        , aborted_(false)
+    {
+        node_->iface()->dispatcher()->register_handler(
+            &eventIdentifyGlobalHandler_, Defs::MTI_EVENTS_IDENTIFY_GLOBAL,
+            Defs::MTI_EXACT);
+    }
+
+    /// Destructor.
+    ~EventIdentifyGlobal()
+    {
+        node_->iface()->dispatcher()->unregister_handler(
+            &eventIdentifyGlobalHandler_, Defs::MTI_EVENTS_IDENTIFY_GLOBAL,
+            Defs::MTI_EXACT);
+    }
+
+    /// Arm the EventIdentifyGlobal request. Multi-thread safe.
+    void arm()
+    {
+        AtomicHolder h(this);
+        if (is_terminated())
+        {
+            aborted_ = false;
+            start_flow(STATE(entry));
+        }
+    }
+
+private:
+    /// Callback upon receiving a Defs::MTI_EVENTS_IDENTIFY_GLOBAL message.
+    /// @param msg unused
+    void event_identify_global(Buffer<GenMessage> *msg)
+    {
+        AtomicHolder h(this);
+        aborted_ = true;
+    }
+
+    /// Entry/reset point into state machine.
+    Action entry()
+    {
+        // get a pseudo random number between 1.500 and 2.011 seconds
+        long long timeout_msec = 1500 + (node_->node_id() & 0x1FF);
```

Comment:
we should try to hash the node ID to do the randomness. The reason for it is that some manufacturers sell hardware in sequential order and mixing bits would be necessary to spread that difference to higher bits of the random selection.
I propose mixing bits with simple arithmetics like this:

```
// This code mixes the bits in the node ID to a hash value in 0 to 511.
uint64_t h = node_->node_id() * 0x1cf9a66d;
uint32_t hh = h ^ (h >> 32);
hh = hh ^ (hh >> 12) ^ (hh >> 24);
timeout_msec = 1500 + (hh & 511);
```

Reply:
Updated/fixed.